libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs

```diff
@@ -116,14 +116,14 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
             return MessageFactory.Text(txt, speak, InputHints.ExpectingInput);
         }
 
-        public static Activity List(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
             return List(ToChoices(choices), text, speak, options);
         }
 
-        public static Activity List(List<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<Choice> choiceList, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
-            choices = choices ?? new List<Choice>();
+            var choices = choiceList?.ToList() ?? new List<Choice>();
```

Comment:
So, in the case of this method (there might be others), if we're accepting an `IEnumerable<T>` now... I would think it would behoove us to improve the implementation(s), when possible, such that we don't have to pay the price of the `.ToList()`.


For example, all we really need to do here is enumerate once over the incoming choices. So we can change this method to use a `foreach` instead with a separate `int index` variable that we increment manually each step and, in fact, you can only bother incrementing it inside of the `if(includeNumbers)` branch.

Also, while I'm here looking at this, this should really be changed to use a `StringBuilder` instead of `string` concats to reduce allocations/garbage.

Reply:
I'll take a closer look. I was trying to avoid touching the implementation deeply, but I'll see what I can clean up here.